src/execution_plan/ops/op_project.c

```diff
@@ -79,13 +79,17 @@ Record ProjectConsume(OpBase *opBase) {
 	int rec_idx = 0;
 	for(unsigned short i = 0; i < op->exp_count; i++) {
 		SIValue v = AR_EXP_Evaluate(op->exps[i], r);
+		// Persisting a value is only necessary here if it belongs to the Record 'r'; can be improved.
+		SIValue_Persist(&v);
 		Record_Add(projection, rec_idx, v);
 		rec_idx++;
 	}
 
 	// Project Order expressions.
 	for(unsigned short i = 0; i < op->order_exp_count; i++) {
 		SIValue v = AR_EXP_Evaluate(op->order_exps[i], r);
+		// Persisting a value is only necessary here if it belongs to the Record 'r'; can be improved.
+		SIValue_Persist(&v);
```

Comment:
will it work with the 
`WITH toupper("a") as x return x`
?
correct me if I'm wrong, but the value in the record will be generated as a result of `toupper` which is a dynamically allocated value and is set as `M_SELF`. The call to persist will not have any effect on it so you'll have two records holding the a copy of the same dynamically allocated value, which they both own. I believe it should be calling to clone

Reply:
In the exact case you describe, `toupper("a")` will use the `AR_EXP_ReduceToScalar` optimization, leading to an AR_ExpNode holding one M_SELF SIValue. When AR_ExpEvaluate is called by the first Project op, a volatile copy will be returned.

In a case like:
`MATCH (a) WITH toupper(a.name) as x return x`
The first Project's call to evaluate returns a self-owned value, but that value does not populate its input Record `r`, only its `projection`. The second Project's AR_ExpNode is just the identifier `x`, which causes a volatile to be returned.